### PR TITLE
Add rank(method="random"); closes #1540

### LINKED ISSUE #1540: Add rank(method="random").
#### Describe your feature request

As of https://github.com/pola-rs/polars/pull/1222 the rank function exists with `min, max, avg, dense, ordinal` as methods.
It would be nice to have `random` too to break ties (where values between min and max are chosen in a random order).

https://stat.ethz.ch/R-manual/R-patched/library/base/html/rank.html

### PR DESCRIPTION
**Pull request by ghuls**

### CODE REVIEW

File: polars/polars-core/src/chunked_array/ops/unique/rank.rs [hunk 1]
```diff
@@ -57,6 +61,59 @@ pub(crate) fn rank(s: &Series, method: RankMethod) -> Series {
...
+                }
+            });
+            // Close last slice (if there where nulls in the original series, they will always be in the last slice).
+            ties_indices.push(len);
+
+            let mut sort_idx = sort_idx.to_vec();
+
+            let rng = &mut thread_rng();
+
+            // Shuffle sort_idx positions which point to ties in the original series.
```
Comment (ghuls) [inline polars/polars-core/src/chunked_array/ops/unique/rank.rs:93]: @ritchie46: Can this be written as an iterator?

File: polars/polars-core/src/chunked_array/ops/unique/rank.rs [hunk 2]
```diff
@@ -52,12 +56,65 @@ pub(crate) fn rank(s: &Series, method: RankMethod) -> Series {
...
-    let inv_ca = UInt32Chunked::new_from_aligned_vec(s.name(), inv);
 
     use RankMethod::*;
     match method {
-        Ordinal => inv_ca.into_series(),
+        Ordinal => {
+            let inv_ca = UInt32Chunked::new_from_aligned_vec(s.name(), inv);
+            inv_ca.into_series()
+        }
+        Random => {
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/unique/rank.rs:66]: Can we add a `[cfg(feature="random")]` on this branch? Then we can still use `rank` if we don't wnat to compile the `random` feature.

File: polars/polars-core/src/chunked_array/ops/unique/rank.rs (hunk 1 above)
Comment (ritchie46) [reply polars/polars-core/src/chunked_array/ops/unique/rank.rs:95]: I think so, but as you shuffle `sort_idx` which is outside of the loop, it doesn't really matter IMO.

File: polars/polars-core/src/chunked_array/ops/unique/rank.rs [hunk 3]
```diff
@@ -52,12 +56,65 @@ pub(crate) fn rank(s: &Series, method: RankMethod) -> Series {
...
+            let mut sort_idx = sort_idx.to_vec();
+
+            let rng = &mut thread_rng();
+
+            // Shuffle sort_idx positions which point to ties in the original series.
+            for i in 0..(ties_indices.len() - 1) as usize {
+                let ties_index_start = ties_indices[i];
+                let ties_index_end = ties_indices[i + 1];
+                if ties_index_end - ties_index_start > 1 {
+                    let _ = &sort_idx[ties_index_start..ties_index_end].shuffle(rng);
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/unique/rank.rs:100]: Does this return a Result? If so, let's unwrap, then we panic if someting goes wrong.

File: polars/polars-core/src/chunked_array/ops/unique/rank.rs [hunk 4]
```diff
@@ -1,12 +1,16 @@
 use crate::prelude::*;
 
+use rand::prelude::SliceRandom;
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/unique/rank.rs:3]: If we conditionally compile, these can be imported in the branch.

File: polars/polars-core/src/chunked_array/ops/unique/rank.rs (hunk 2 above)
Comment (ghuls) [reply polars/polars-core/src/chunked_array/ops/unique/rank.rs:78]: added.

File: polars/polars-core/src/chunked_array/ops/unique/rank.rs (hunk 4 above)
Comment (ghuls) [reply polars/polars-core/src/chunked_array/ops/unique/rank.rs:4]: So where should it be added?

File: polars/polars-core/src/chunked_array/ops/unique/rank.rs (hunk 3 above)
Comment (ghuls) [reply polars/polars-core/src/chunked_array/ops/unique/rank.rs:112]: It does not return a result:
```
error[E0599]: no method named `unwrap` found for unit type `()` in the current scope
   --> polars/polars-core/src/chunked_array/ops/unique/rank.rs:112:78
    |
112 |                     &sort_idx[ties_index_start..ties_index_end].shuffle(rng).unwrap();
    |                                                                              ^^^^^^ method not found in `()`
```

https://docs.rs/rand/0.8.4/rand/seq/trait.SliceRandom.html#tymethod.shuffle

File: polars/polars-core/src/chunked_array/ops/unique/rank.rs (hunk 4 above)
Comment (ritchie46) [reply polars/polars-core/src/chunked_array/ops/unique/rank.rs:4]: This is also fine. :+1:

File: polars/polars-core/src/chunked_array/ops/unique/rank.rs (hunk 3 above)
Comment (ritchie46) [reply polars/polars-core/src/chunked_array/ops/unique/rank.rs:112]: Ok, then I don't think we have to bind it to `let _`. It made me wonder why we did that. :P

Comment (ghuls) [reply polars/polars-core/src/chunked_array/ops/unique/rank.rs:112]: without `let _`: ```rust
warning: unused borrow that must be used
   --> polars/polars-core/src/chunked_array/ops/unique/rank.rs:113:21
    |
113 |                     &sort_idx[ties_index_start..ties_index_end].shuffle(rng);
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the borrow produces a value
    |
    = note: `#[warn(unused_must_use)]` on by default
help: use `let _ = ...` to ignore the resulting value
    |
113 |                     let _ = &sort_idx[ties_index_start..ties_index_end].shuffle(rng);
    |                     +++++++
```

Comment (ritchie46) [reply polars/polars-core/src/chunked_array/ops/unique/rank.rs:112]: I think we must borrow `&mut`. :thinking:

Comment (ghuls) [reply polars/polars-core/src/chunked_array/ops/unique/rank.rs:112]: `sort_idx[ties_index_start..ties_index_end].shuffle(rng);` seems to work.

**Comment by ritchie46** [comment]
Thanks for helping on this one!

**Pull request merged**